### ISSUE TITLE
Metadata caching between "hive-style" catalogs

### ISSUE THREAD

**Issue by wrb2** [trinodb/trino#15656]
When using multiple "hive-style" catalogs connected to same Hive Metastore or Glue with redirection (e.g. running Delta connector and Hive connector and redirecting between them), if metadata caching is enabled, the caches are not synced.

This is a problem in scenarios where table or partition is created using one catalog and read using another, or dropped in one catalog but the other still sees it.

It especially impacts a use case of changing table format (e.g. migrating from Hive to Delta) by dropping table in one catalog and recreating it in the other catalog. That gets Trino into a state where re-running DROP fails because "table doesn't exist" and trying to create it also fails because "table already exists" and the user needs to either flush cache manually or restart Trino to get catalogs back into consistent state.

Currently this impacts Hive and Delta catalogs, because Iceberg catalog doesn't implement metadata caching, but if it ever does, it might be impacted as well.

**Comment by findinpath** [comment]
> This is a problem in scenarios where table or partition is created using one catalog and read using another, or dropped in one catalog but the other still sees it.

Do note that the catalog which is responsible for a table (depending on its type - e.g. : hive / delta / iceberg / hudi) will be the catalog performing operations on this table.


Do you have a reproduction scenario for the potential problem you are speaking about?

**Comment by wrb2** [comment]
> Do you have a reproduction scenario for the potential problem you are speaking about?

Let's say I have catalog of type hive, called `prod`. I want to move from hive to delta, so I configure delta catalog and call it `prod`, rename the hive catalog to `prod_legacy` and setup redirection from `prod` to `prod_legacy`.

As long as everyone is just reading, users can keep using `prod` catalog, applications that use data are not impacted, the users won't even know - new tables created in `prod` catalog are now delta tables but old tables still work. So far so good.

This breaks when someone tries to drop and recreate a table (for example by running dbt).

Running `drop table prod.blah` on a hive table will redirect to `drop table prod_legacy.blah` and successfully drop the table. However, if metadata caching is enabled, the table will stay cached in `prod`. So subsequent `create table prod.blah` will fail saying that table already exists. But running `drop table prod.blah` again will redirect to `drop table prod_legacy.blah` - and it will fail because table doesn't exist. At this point the only way to fix it is to flush cache manually or restart Trino.

I get it might sound a bit contrived but this use case of gradually switching from hive table format to a smarter one like delta (or iceberg once it implements cashing) without impacting users by switching catalog names and using redirection is really valuable, compared to say, telling hundreds of users "can you please use prod_delta now" or something.

And of course, larger installations of hive or especially glue catalog are really impossible without having the caching enabled, so turning it off is not really realistic.

**Comment by findinpath** [comment]
@wrb2  thank you for taking the time to describe in detail the potential bug in the table/catalog redirection mechanism.


I tried the scenario you previously exposed on my machine by doing:

```
testing/bin/ptl env up --environment singlenode-delta-lake-oss --without-trino
```

On Trino side I kept only `hive.properties` and `delta.properties` and renamed them accordingly to your description to `prod_legacy.properties` and `prod.properties`.

I created two tables:

- `prod_legacy.tiny.hivetable1`
- `prod.tiny.deltatable1`

Below is the listing of the commands performed based on your description:

```
trino> show tables in prod.tiny;
    Table    
-------------
 deltatable1 
 hivetable1  
(2 rows)

Query 20230111_091904_00009_m4a37, FINISHED, 1 node
http://localhost:8080/ui/query.html?20230111_091904_00009_m4a37
Splits: 11 total, 11 done (100.00%)
CPU Time: 0.0s total,    60 rows/s, 1.45KB/s, 36% active
Per Node: 0.1 parallelism,     7 rows/s,   179B/s
Parallelism: 0.1
Peak Memory: 796B
0.27 [2 rows, 49B] [7 rows/s, 179B/s]

trino> drop table prod_legacy.tiny.hivetable1;
DROP TABLE
trino> create table prod.tiny.hivetable1 (c1 integer, c2 integer);
CREATE TABLE
```

The caching mechanism present in the Delta Lake connector is characteristic to table metadata files and not to tables.

The table or catalog redirection mechanism reaches the metastore, verifies what kind of table is dealing with and redirects the flow to the appropriate connector. When dropping a hive table, the flow will be redirected towards the Hive connector (`prod_legacy in your example) leaving no footprint at all on the Delta Lake connector (`prod` in your example).

Even though I find perfectly valid the migration scenario you have exposed, I still fail to understand how to reproduce the behaviour you exposed. Can you please help out with more details on how to eventually reproduce the eventual problem?

**Comment by wrb2** [comment]
Right, I didn't explain it properly - what I meant by metadata caching is for **both** of the catalogs to have configured something like this:

```properties
hive.metastore-refresh-interval=30m
hive.metastore-cache-ttl=360m
hive.metastore-cache-maximum-size=80000
```

What I did was

```sql
create table prod_legacy.tiny.table1 as select 1 as x; -- create hive table
select * from system.jdbc.tables; -- table should be visible in both
select * from prod.tiny.table1; -- works due to redirect
drop table prod.tiny.table1; -- works due to redirect
select * from system.jdbc.tables; -- for me, table still shows up in prod, but the prod_legacy one is gone
create table prod.tiny.table1 as select 1 as x; -- fails
```

I found this behavior on Trino 402 with Glue Catalog - I'm not sure if the metadata cache handling changed between version or if it might be different between Hive and Glue?

**Comment by findinpath** [comment]
I'm currently testing on `master` Trino `406`.

I took the extra properties for the `hive` connector (or `prod_legacy` in your example) and tried them out.


I tried out your example queries and received the following results:

```
create table prod_legacy.tiny.table1 as select 1 as x; -- create hive table
select * from system.jdbc.tables; -- table should be visible in both
select * from prod.tiny.table1; -- works due to redirect
drop table prod.tiny.table1; -- works due to redirect
```

```
trino:default> drop table prod.tiny.table1; 
DROP TABLE
trino:default> select * from system.jdbc.tables where table_name='table1';
 table_cat | table_schem | table_name | table_type | remarks | type_cat | type_schem | type_name | self_referencing_col_name |>
-----------+-------------+------------+------------+---------+----------+------------+-----------+---------------------------+>
(0 rows)

Query 20230112_090218_00004_8rccr, FINISHED, 1 node
http://localhost:8080/ui/query.html?20230112_090218_00004_8rccr
Splits: 1 total, 1 done (100.00%)
CPU Time: 0.0s total, 7.14K rows/s,  236KB/s, 7% active
Per Node: 0.0 parallelism,   234 rows/s, 7.76KB/s
Parallelism: 0.0
Peak Memory: 64B
0.21 [50 rows, 1.65KB] [234 rows/s, 7.76KB/s]

trino:default> create table prod.tiny.table1 as select 1 as x;
CREATE TABLE: 1 row

Query 20230112_090246_00005_8rccr, FINISHED, 1 node
http://localhost:8080/ui/query.html?20230112_090246_00005_8rccr
Splits: 26 total, 26 done (100.00%)
CPU Time: 0.6s total,     0 rows/s,     0B/s, 68% active
Per Node: 0.4 parallelism,     0 rows/s,     0B/s
Parallelism: 0.4
Peak Memory: 1.57KB
1.54 [0 rows, 0B] [0 rows/s, 0B/s]

trino:default> select * from system.jdbc.tables where table_name='table1';
 table_cat | table_schem | table_name | table_type | remarks | type_cat | type_schem | type_name | self_referencing_col_name |>
-----------+-------------+------------+------------+---------+----------+------------+-----------+---------------------------+>
 prod      | tiny        | table1     | TABLE      | NULL    | NULL     | NULL       | NULL      | NULL                      |>
(1 row)
```


Not seeing the `prod_legacy.tiny.table` in `system.jdbc.tables` can be explained through the fact that the `tiny` schema is being cached in the  caching layer of `prod_legacy` hive catalog.

Nevertheless, `SELECT` statement still works on both catalogs because it doesn't go through the caching layer.

```
trino:default> select * from prod_legacy.tiny.table1;
 x 
---
 1 
(1 row)

Query 20230112_091609_00010_8rccr, FINISHED, 1 node
http://localhost:8080/ui/query.html?20230112_091609_00010_8rccr
Splits: 1 total, 1 done (100.00%)
CPU Time: 0.1s total,    17 rows/s,  3.7KB/s, 64% active
Per Node: 0.2 parallelism,     3 rows/s,   681B/s
Parallelism: 0.2
Peak Memory: 140B
0.32 [1 rows, 216B] [3 rows/s, 681B/s]

trino:default> select * from prod.tiny.table1;
 x 
---
 1 
(1 row)
```

I couldn't reproduce though the `CREATE TABLE` failure.
